test/lit/ctor-eval/gc-cycle.wast

```diff
@@ -0,0 +1,1297 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.
+;; RUN: foreach %s %t wasm-ctor-eval --ctors=test --kept-exports=test --quiet -all -S -o - | filecheck %s
+
+(module
+ ;; CHECK:      (type $A (struct (field (mut (ref null $A))) (field i32)))
+ (type $A (struct (field (mut (ref null $A))) (field i32)))
+
+ ;; CHECK:      (type $none_=>_none (func))
+
+ ;; CHECK:      (type $none_=>_i32 (func (result i32)))
+
+ ;; CHECK:      (global $ctor-eval$global_3 (ref $A) (struct.new $A
+ ;; CHECK-NEXT:  (ref.null none)
+ ;; CHECK-NEXT:  (i32.const 42)
+ ;; CHECK-NEXT: ))
+
+ ;; CHECK:      (global $a (mut (ref null $A)) (global.get $ctor-eval$global_3))
+ (global $a (mut (ref null $A)) (ref.null $A))
+
+ (func "test"
+  (local $a (ref $A))
+  ;; This generates a self-cycle where the global $a's ref field points to
+  ;; itself. To handle this, wasm-ctor-eval will emit a new global with a null
+  ;; in the ref field, and add a start function that adds the self-reference.
+  (global.set $a
+   (local.tee $a
+    (struct.new $A
+     (ref.null $A)
+     (i32.const 42)
+    )
+   )
+  )
+  (struct.set $A 0
+   (local.get $a)
+   (local.get $a)
+  )
+ )
+
+ ;; CHECK:      (export "test" (func $0_3))
+
+ ;; CHECK:      (export "keepalive" (func $keepalive))
+
+ ;; CHECK:      (start $start)
+
+ ;; CHECK:      (func $keepalive (type $none_=>_i32) (result i32)
+ ;; CHECK-NEXT:  (struct.get $A 1
+ ;; CHECK-NEXT:   (struct.get $A 0
+ ;; CHECK-NEXT:    (global.get $a)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $keepalive (export "keepalive") (result i32)
+  ;; Getting $A.0.1 (reading from the reference in the global's first field)
+  ;; checks that we have a proper reference there. If we could do --fuzz-exec
+  ;; here we could validate that (but atm we can't use --output=fuzz-exec at the
+  ;; same time as --all-items in the update_lit_checks.py note).
+  (struct.get $A 1
+   (struct.get $A 0
+    (global.get $a)
+   )
+  )
+ )
+)
+
+;; CHECK:      (func $start (type $none_=>_none)
+;; CHECK-NEXT:  (struct.set $A 0
+;; CHECK-NEXT:   (global.get $ctor-eval$global_3)
+;; CHECK-NEXT:   (global.get $ctor-eval$global_3)
+;; CHECK-NEXT:  )
+;; CHECK-NEXT: )
+
+;; CHECK:      (func $0_3 (type $none_=>_none)
```

Comment:
I believe you wrote before that naming the function `$test` won't make wasm-ctor-eval preserve its name. Why does function `test` get renamed to `0_3`? Shouldn't wasm-ctor-eval be able to preserve the function names?

Reply:
It can at least keep the base name, I guess, even if it adds a prefix... I added base names now.

It adds a prefix because in general an export might have another use. So we copy the function, modify that, and make the export use that one (so that other uses still use the original, unmodified function). As a result the export points to a function with a different name than before.